### ISSUE TITLE
ERC1238: Non-transferrable Non-Fungible Tokens (NTT)

### ISSUE THREAD

**Issue by nicola** [ethereum/EIPs#1238]
```
eip: 1238
title: Non-transferrable Non-Fungible Tokens (badges)
author: Nicola Greco (and future others)
status: WIP
category: ERC
created: 2018-07-20
```

# Non-transferable tokens

## Simple Summary
A *badge* is a token that once assigned it cannot be transferred. Badges can be accumulated through time and put at stake. Simply speaking, badges are statements about a public key, they can be quantitative (e.g. reputation, experience) or qualitative (badges, titles).

## Abstract
The *Non-transferrable token* standard defines a set of standard APIs allowing the identification of statements (called *badges*) attributed to a public key, such that different dapps and smart contract can use to filter users or to provide user with different badges different experiences. More importantly, this standard defines a way for which users can put their badges at stake. Badges cannot be transferred but can be lost (after staking) or can expire.

## Motivation
The inspiration for designing badges comes from a conversation with Andrew Miller. The idea is to have a token that once assigned it cannot be transferred (like reputation) and that it can be used by websites, or contracts to make me perform some actions. For example, if a user accumulates paper submissions at conferences, then they can use their paper badges to request grants. It's important that they can never share these badges.

This is the equivalent of a variety of other use cases
- Accumulating badges: degrees from academic institutions, paper publications for conferences, badges that allow for access in a building or in a special chat.
- Experience points: points accumulated in a game, points accumulated by being honest in a decentralized system for some time
- Statements: more broadly, any statement that is issued or signed by a contract, a dao, a single user that requires to be on-chain
- Subscription: badges can represent the validity of a paid subscription

## Requirements for specification
- **Quantity**: a badge can have the form of 1-many, meaning that a badge can represent a single value (Degree from CS, 1), or can a quantity (Experience in Fortnite, 2000)
- **Stake**: it should be simple to put at stake contingent on a smart contract execution (however, it is understandable that this could be done by a different standard)
- **Backwards compatibility**: This standard should have similar apis to existing standards (e.g. #20, #721), I makes sense for an NFT viewer to also view badges
- **Off-chain tokens**: This specification should work also for tokens that are not issued on chain (say by a third party signer) and can be presented on chain only when necessary (it's understandable that this is an edge case and could fall off the spec, but it's an interesting one!)

## Next steps
There are two main steps here:
- Specification
- Implementation

However, I will have very little time to actually bring this to life. It would be awesome if any contributor here feels strong and empowered to take over this standard and make this their own. I started this while thinking on a fun project for academia (namely called [research coin](https://nicola.io/research-coin/2017)), maybe someone wants to help there!

**Comment by adibas03** [comment]
I think a way to escape the address restriction is to utilize the proxy implementation #121 such that the proxy has the badge and such the owner can have access with any of the owned addresses

**Comment by chiro-hiro** [comment]
@nicola will the reputation be decayed overtime? 

The same thing happens in real world :)

**Comment by Arachnid** [comment]
@adibas03 That demonstrates how pointless preventing transfers would be - it's easily evaded, and so just makes life more difficult for users.

**Comment by nicola** [comment]
@chiro-hiro it would be up to you

@Arachnid I agree and disagree, the identity would be an ID where you can decide the set of keys that own that ID, you can't transfer badges across IDs.. For example, when playing in a game, you only have one ID per player and you only have some experience. In other words, I believe there is something useful here.

note: re-issuing is still a valid option for some usecases

**Comment by adibas03** [comment]
@Arachnid I see your point about the `non-transferrale` attribute. Similar to game credits, which are not natively transferrable, but the accounts with the credits and repuation can be sold. So, the proxy/Identity accounts can be transferred in a way.
@nicola I think @Arachnid has a point here.
Also, could the badges be an example of a claim as described in #735

**Comment by MicahZoltu** [comment]
There is a big difference between transferring individual badges and transferring all badges associated with a key.  I think there is noue in allowing users to transfer all of their badges to a new key, while also preventing individual badges from being transferred away.

One can think of the set of badges as identifying a user, and you cannot split the user into parts and have it be the same user, but you can transfer a user to a new private key.

**Comment by vongohren** [comment]
What differences do you see between badges and claims? 
Ref: https://medium.com/uport/erc1056-erc780-an-open-identity-and-claims-protocol-for-ethereum-aef7207bc744

Both #735 and uPort uses the wording of claims.

**Comment by nicola** [comment]
#735 looks very similar, if not identical to that, I will reach out to them.

**Comment by PhABC** [comment]
I worked on some "badge" token stuff to quantify participation events for fun and also opted for "transfer all or none scheme" as @MicahZoltu proposes, which I think is good enough for the intent here.

**Comment by jcksncllwy** [comment]
@vongohren A significant difference between #735 claims and #1238 badges is information ownership. In #735 the Claim Holder owns any claims made about themselves. The problem with this is that there is no way for a Claim Issuer to revoke or alter a claim once it has been issued. While #735 does specify a `removeClaim` method, a malicious implementation could simply ignore that method call, because _they own the claim_.

Imagine that SafeEmploy™, a background checking company, issues a claim about Timmy. The claim states that Timmy has never been convicted of any felonies. Timmy makes some bad decisions, and now that claim is no longer true. SafeEmploy™ executes `removeClaim`, but Timmy's #735 contract just ignores it, because Timmy wants to stay employed (and is crypto-clever).

#1238 badges do not have this problem. Ownership of a badge/claim is entirely determined by the contract **issuing** the badges, not the one **receiving** them. The issuer is free to remove or change those badges as they see fit.

**Comment by conejoninja** [comment]
@jcksncllwy In #725 (along with #735) The claim issuer could remove the key (from their own identity) that signed the claim to revoke it.

**Comment by OFRBG** [comment]
Reputation system are likely to be abused. Cutting the explanation and straight to the example:

> A major org around Ethereum decides to screen users. The screening process may only be performed by said org, since they set the rules for their own DApp. Being the first mover with a household name.
> 
> Since Ethereum doesn't have truly private methods and attributes, other DApp groups can easily read the badges other users have. If the major org with first mover advantage creates its own version of crypto-TSA, other groups will simply save time and resources, piggybacking on what's available.
> 
> If the problem isn't already blatantly obvious, think about the reputation bottleneck this causes: the major org or orgs that work as a crypto-TSA become gatekeepers for the whole system. We know from history that some things we do aren't the best, but they stuck because they were the most comfortable. When going from a reputation-less system into a reputation-based system, whatever comes first is most likely to stay. 

On a trustless network every address should be no different from any other. I think a measured reputation system goes against what crypto should be. In other words, creating an issued-reputation ERC is opening Pandora's Box.

**Comment by adibas03** [comment]
@OFRBG You are right about the danger of Reputation, but as you mentioned, the implementation plays a big role.All or null transfer-ability as suggested by @MicahZoltu still seems the best possible option.
Also, the two main preferable  implementation I believe are (I am using `claim` t refer to both `claims`  #735 and `badges` #1238),
- Claim registry contract which allows contract owner create claims on address which can be approved or rejected by the address owner. The state of the contract concerning the claim can be regarded as final, as the contract gets more approved claimed, it reputation increases, and that is when things can get interesting
- Claimant makes a claim on an address, which is approved by the address by publishing the claim identifier (needs a standard). The claimant can at any time withdraw the claim and the address can at any time have the claim identifier information unpublished, thereby voiding the claim. In this case, both Claimant (contract) and address (Identity contract #725 ) both have to publish the same information to be considered approved- This is a whole lot more complicated, but I think quenches mot of the worries attached to the security concerns.

**Comment by snaketh4xor** [comment]
nice eip @nicola 
I just got some free time from my busy college so I decided to contribute to decentralize web.
so here is example code with non transferable functionality implemented.

Badge is **NON transferable token** issued by owner to a particular identity.

badges are issued only once to new identities.
to issue new badges new contracts have to be deployed.
there can be a new function introduced to remove a identity from transfer restriction so it can be issued more badge tokens(which is of no use) but then it will have drawbacks as described in second way further below.

There were two ways to implement it,

first by preventing transfer for each identity which uses lot of gas as each identity is stored on blockchain.
secondly by freezing transfer function of contract,its gas efficient but it had few drawbacks
consider owner unfreezed transfer to issue new badge token for particular new identity,during this time any existing token holder could transfer token,onlyOwner modifier on transfer is of no use as owner would be able to transfer a token between identities hence owner can be bribed by token holders to transfer their tokens to another token holder but we have to prevent that.
so first way is best way plus owner can set his address as target and so no one will be able transfer any token in future not even owner hence no new token issued.

owner here can be
single address or multi-sig address contract controlled by single individual or group->centralized
DAO->decentralized
public by anyone->truly decentralized

Repution system is not implemented as secure,safe reputation system on chain is not proposed yet,all current iterations of reputation system on chain are exploitable,defining it into a standard will be like making a bug a standard.

off chain tokens can be created thanks to create2 but it would be better to use it in universal login.
so a proxy identity contract will be created for a user to which owner will issue a badge or non transferable token and it will be controlled by set of allowed keys.

**->code starts here**

/*// ----------------------------------------------------------------------------
// ERC Token Standard #20 Interface extended to Non transferable token(Badge) standard
// https://github.com/ethereum/EIPs/blob/master/EIPS/eip-20.md
// ----------------------------------------------------------------------------
contract ERC20Interface {
    function totalSupply() public view returns (uint);
    function balanceOf(address tokenOwner) public view returns (uint balance);
    function allowance(address tokenOwner, address spender) public view returns (uint remaining);
    function freezeAccount(address target,bool freeze) public returns (bool success);
    function transfer(address to, uint tokens) public returns (bool success);
    function approve(address spender, uint tokens) public returns (bool success);
    function transferFrom(address from, address to, uint tokens) public returns (bool success);

    event Transfer(address indexed from, address indexed to, uint tokens);
    event Approval(address indexed tokenOwner, address indexed spender, uint tokens);
    event NonTransferable(address target, bool frozen);

*/

pragma solidity ^0.5.0;

// ----------------------------------------------------------------------------
// 'FIXED' 'Example Fixed Supply Non Transferable Token' token contract
//
// Symbol      : FIXED
// Name        : Example Fixed Supply Non Transferable Token
// Total supply: 1,000,000.000000000000000000
// Decimals    : 18
//
// Enjoy.
//
// ERC-20 standard taken from (c) BokkyPooBah / Bok Consulting Pty Ltd 2018. The MIT Licence.
//thanks for your contribution to community
//extended for non transferable token by Snaketh4xor
// ----------------------------------------------------------------------------


// ----------------------------------------------------------------------------
// Safe maths
// ----------------------------------------------------------------------------
library SafeMath {
    function add(uint a, uint b) internal pure returns (uint c) {
        c = a + b;
        require(c >= a);
    }
    function sub(uint a, uint b) internal pure returns (uint c) {
        require(b <= a);
        c = a - b;
    }
    function mul(uint a, uint b) internal pure returns (uint c) {
        c = a * b;
        require(a == 0 || c / a == b);
    }
    function div(uint a, uint b) internal pure returns (uint c) {
        require(b > 0);
        c = a / b;
    }
}


// ----------------------------------------------------------------------------
// ERC Token Standard #20 Interface
// https://github.com/ethereum/EIPs/blob/master/EIPS/eip-20.md
// ----------------------------------------------------------------------------
contract ERC20Interface {
    function totalSupply() public view returns (uint);
    function balanceOf(address tokenOwner) public view returns (uint balance);
    function allowance(address tokenOwner, address spender) public view returns (uint remaining);
    function transfer(address to, uint tokens) public returns (bool success);
    function approve(address spender, uint tokens) public returns (bool success);
    function transferFrom(address from, address to, uint tokens) public returns (bool success);

    event Transfer(address indexed from, address indexed to, uint tokens);
    event Approval(address indexed tokenOwner, address indexed spender, uint tokens);
}


// ----------------------------------------------------------------------------
// Contract function to receive approval and execute function in one call
//
// Borrowed from MiniMeToken
// ----------------------------------------------------------------------------
contract ApproveAndCallFallBack {
    function receiveApproval(address from, uint256 tokens, address token, bytes memory data) public;
}


// ----------------------------------------------------------------------------
// Owned contract
// ----------------------------------------------------------------------------
contract Owned {
    address public owner;
    address public newOwner;

    event OwnershipTransferred(address indexed _from, address indexed _to);

    constructor() public {
        owner = msg.sender;
    }

    modifier onlyOwner {
        require(msg.sender == owner);
        _;
    }

    function transferOwnership(address _newOwner) public onlyOwner {
        newOwner = _newOwner;
    }
    function acceptOwnership() public {
        require(msg.sender == newOwner);
        emit OwnershipTransferred(owner, newOwner);
        owner = newOwner;
        newOwner = address(0);
    }
}


// ----------------------------------------------------------------------------
// ERC20 Token, with the addition of symbol, name and decimals and a
// fixed supply extended to Non transferable token 
// ----------------------------------------------------------------------------
contract FixedSupplyToken is ERC20Interface, Owned {
    using SafeMath for uint;

    string public symbol;
    string public  name;
    uint8 public decimals;
    uint _totalSupply;

    mapping(address => uint) balances;
    mapping(address => mapping(address => uint)) allowed;
    mapping(address => bool) public frozenAccount;
    
    event NonTransferable(address target, bool frozen);

    // ------------------------------------------------------------------------
    // Constructor
    // ------------------------------------------------------------------------
    constructor() public {
        symbol = "FIXED";
        name = "Example Fixed Supply Token";
        decimals = 18;
        _totalSupply = 1000000 * 10**uint(decimals);
        balances[owner] = _totalSupply;
        emit Transfer(address(0), owner, _totalSupply);
    }


    // ------------------------------------------------------------------------
    // Total supply
    // ------------------------------------------------------------------------
    function totalSupply() public view returns (uint) {
        return _totalSupply.sub(balances[address(0)]);
    }


    // ------------------------------------------------------------------------
    // Get the token balance for account `tokenOwner`
    // ------------------------------------------------------------------------
    function balanceOf(address tokenOwner) public view returns (uint balance) {
        return balances[tokenOwner];
    }

    //stops transfer of tokens for identity.
    function freezeAccount(address target, bool freeze) onlyOwner public returns (bool success) {
        frozenAccount[target] = freeze;
        emit NonTransferable(target, freeze);
        return true;
    }
    // ------------------------------------------------------------------------
    // Transfer the balance from token owner's account to `to` account
    // - Owner's account must have sufficient balance to transfer
    // - 0 value transfers are allowed
    // ------------------------------------------------------------------------
    function transfer(address to, uint tokens) onlyOwner public returns (bool success) {
        require(!frozenAccount[to]);
        require(!frozenAccount[msg.sender]);
        balances[msg.sender] = balances[msg.sender].sub(tokens);
        balances[to] = balances[to].add(tokens);
        freezeAccount(to, true);
        emit Transfer(msg.sender, to, tokens);
        return true;
    }


    // ------------------------------------------------------------------------
    // Token owner can approve for `spender` to transferFrom(...) `tokens`
    // from the token owner's account
    //
    // https://github.com/ethereum/EIPs/blob/master/EIPS/eip-20-token-standard.md
    // recommends that there are no checks for the approval double-spend attack
    // as this should be implemented in user interfaces
    // ------------------------------------------------------------------------
    function approve(address spender, uint tokens) onlyOwner public returns (bool success) {
        allowed[msg.sender][spender] = tokens;
        emit Approval(msg.sender, spender, tokens);
        return true;
    }


    // ------------------------------------------------------------------------
    // Transfer `tokens` from the `from` account to the `to` account
    //
    // The calling account must already have sufficient tokens approve(...)-d
    // for spending from the `from` account and
    // - From account must have sufficient balance to transfer
    // - Spender must have sufficient allowance to transfer
    // - 0 value transfers are allowed
    // ------------------------------------------------------------------------
    function transferFrom(address from, address to, uint tokens) onlyOwner public returns (bool success) {
        require(!frozenAccount[to]);
        require(!frozenAccount[from]);
        balances[from] = balances[from].sub(tokens);
        allowed[from][msg.sender] = allowed[from][msg.sender].sub(tokens);
        balances[to] = balances[to].add(tokens);
        freezeAccount(to, true);
        emit Transfer(from, to, tokens);
        return true;
    }

    // ------------------------------------------------------------------------
    // Returns the amount of tokens approved by the owner that can be
    // transferred to the spender's account
    // ------------------------------------------------------------------------
    function allowance(address tokenOwner, address spender) public view returns (uint remaining) {
        return allowed[tokenOwner][spender];
    }


    // ------------------------------------------------------------------------
    // Token owner can approve for `spender` to transferFrom(...) `tokens`
    // from the token owner's account. The `spender` contract function
    // `receiveApproval(...)` is then executed
    // ------------------------------------------------------------------------
    function approveAndCall(address spender, uint tokens, bytes memory data) onlyOwner public returns (bool success) {
        allowed[msg.sender][spender] = tokens;
        emit Approval(msg.sender, spender, tokens);
        ApproveAndCallFallBack(spender).receiveApproval(msg.sender, tokens, address(this), data);
        return true;
    }


    // ------------------------------------------------------------------------
    // Don't accept ETH
    // ------------------------------------------------------------------------
    function () external payable {
        revert();
    }


    // ------------------------------------------------------------------------
    // Owner can transfer out any accidentally sent ERC20 tokens
    // ------------------------------------------------------------------------
    function transferAnyERC20Token(address tokenAddress, uint tokens) public onlyOwner returns (bool success) {
        return ERC20Interface(tokenAddress).transfer(owner, tokens);
    }
}

**Comment by winnal** [comment]
I was thinking of a use-case of a non-transferable token for the purposes of a founders token, where tokens are bound to the user account, where the user can use the tokens for voting and spending to redeem a benefit, but cannot trade or transfer the token, which would make a founders token sale (FTO) exempt from securities regulation as the token cannot be traded, yet would enable there to be a development fund through a token sale.

**Comment by xuhcc** [comment]
I think that it is sufficient to provide a single method that returns boolean value indicating transferrability of a token. All other details should be left to implementers. Such interface will be compatible with ERC20, ERC721, ERC1155 and other token standards.

Here's an example:

```js
contract ERC1238 is ERC165 {

    bool internal _transfersEnabled;

    /*
     *     bytes4(keccak256('transfersEnabled()')) == 0xbef97c87
     */
    bytes4 private constant _INTERFACE_ID_ERC1238 = 0xbef97c87;

    constructor () public {
        _registerInterface(_INTERFACE_ID_ERC1238);
    }
    
    function transfersEnabled() external view returns (bool) {
        return _transfersEnabled;
    }
}
```

The name `trasfersEnabled` is borrowed from [MiniMe token contract](https://github.com/Giveth/minime/blob/master/contracts/MiniMeToken.sol#L81). It is also used by [Aragon](https://github.com/aragon/aragon-apps/blob/shared-minime-v1.0.2/shared/minime/contracts/MiniMeToken.sol#L101) to designate non-transerrable reputation and membership tokens in DAOs.

Transferrability of ERC1238 token can change depending on some condition, so it allows complex logic like the "all or none" scheme described earlier in this thread.

**Comment by rosspeili** [comment]
I am not sure how viable this contract can be, but having in mind potential future use-cases, I assume it should be abuse-proof. For example, We create a biometric web3 ID by completing a DNA sequencing, we upload this data to a bioinformatics related blockchain like Zenome, ARNA, or Encrypgen. The respective DLT platform could mint NTTs based on a KYC procedure that includes the genomic data. This NTT could be used as a web3 ID, and it can also be used to retrieve a wallet in case of loss of all recovery methods including private key and seed phrase. Eg. I lost my web3 wallet that has a biometric NTT, I perform a new test or request my test from the lab, or blockchain that created the initial contract using real-time KYC. Proving our new or retrieved DNA test, we could retrieve our wallet, again, considering the ERC is secure and truly nontransferable. That for, if once I mint my own biometric ID and manage to transfer it to your wallet, I could easily create a second test to claim access to your wallet that will basically indicate it bears my DNA. NTT must be obsolete. It's nice to have NTTs with expiration data, but biometric IDs could be only one time mint, one time transfer, nonfungible, and not transferable tokens. That's one of the most exciting ERCs and its a shame DeFi dudes are not paying attention.

**Comment by Genobank** [comment]
@rosspeili please review our peer reviewed article.  Happy to collaborate. 

"Privacy Laws, Genomic data and Non-fungible-tokens"
https://jbba.scholasticahq.com/article/13164.pdf

**Comment by Genobank** [comment]
@rosspeili please review our peer reviewed article.  Happy to collaborate. 

"Privacy Laws, Genomic data and Non-fungible-tokens"
https://jbba.scholasticahq.com/article/13164.pdf

**Comment by fulldecent** [comment]
If anybody is interested in an implementation of the concepts discussed in this thread, please see our publication:

https://genobank.io/biosample-permission-token-with-non-fungible-tokens

---

Here we are creating tokens that allow the creator to edit them at any time.

All tokens are created as ERC-721s and are namespaced to the account owner. i.e. bits 1...160 are the address of `msg.sender` and the remaining 96 bits are chosen by the token creator.

Effectively this allows anybody to create badges/tokens and they are not transferred.

---

And in other words, it is creating a public data structure:

mapping (OWNER address => mapping (uint96 => (TOKEN METADATA, ADDRESS)))

**Comment by ra-phael** [comment]
Partly inspired by this issue, @arcalinea and I are working on a reputation service that uses non-transferrable, non-fungible tokens as on-chain proofs of reputation. The token, which we call a “badge”, gets issued to an address that a user wants to link to a centralized service (i.e. Twitter, Github), through our third-party reputation verification service.

We are still playing with this concept, and have created a contract issuing these badges to the [Kovan testnet.](https://kovan.etherscan.io/address/0xb3e404E8BCbc2af1683da73f14A558662346119F) Any feedback is welcome!

At first, we drafted a version that is ERC-721 compliant, but that adds unnecessary overhead to the simple concept of a badge, which does not need approve and transfer functions, or a tokenURI field with associated metadata. The badge contract we have deployed has different requirements. These are the basic things we believe a badge contract implementation needs to have:

- non-transferability of the token
- a clear link to the issuer of the badge
- a “burn” function to revoke badges
- a timestamp at time of issuance

If this type of token becomes useful for applications, it would be great to standardize non-transferrable, non-fungible tokens.

Some contract code is below. The version below does not have a token struct with id and timestamp - doing so makes querying for the time the badge was created easier, but costs more in gas. An alternate version with timestamps can be found [here](https://github.com/ra-phael/badge-token/blob/main/contracts/IBadge.sol).

Version without timestamp

```solidity=
interface IBadge {
    // @dev Emitted when `tokenId` token is minted to `to`, an address.
    event Minted(
        address indexed to,
        bytes32 indexed tokenId,
        uint256 timestamp
    );

    // @dev Emitted when `tokenId` token is burned.
    event Burned(
        address indexed owner,
        bytes32 indexed tokenId,
        uint256 timestamp
    );

    // @dev Returns the badge's name
    function name() external view returns (string memory);

    // @dev Returns the badge's symbol.
    function symbol() external view returns (string memory);

    // @dev Returns the ID of the token owned by `owner`, if it owns one, and 0 otherwise
    function tokenOf(address owner) external view returns (bytes32);

    // @dev Returns the owner of the `tokenId` token.
    function ownerOf(bytes32 tokenId) external view returns (address);
}

contract Badge is IBadge {
    // Badge's name
    string private _name;

    // Badge's symbol
    string private _symbol;

    // Mapping from token ID to owner's address
    mapping(bytes32 => address) private _owners;

    // Mapping from owner's address to token ID
    mapping(address => bytes32) private _tokens;

    constructor(string memory name_, string memory symbol_) {
        _name = name_;
        _symbol = symbol_;
    }

    // Returns the badge's name
    function name() public view virtual override returns (string memory) {
        return _name;
    }

    // Returns the badge's symbol
    function symbol() public view virtual override returns (string memory) {
        return _symbol;
    }

    // Returns the token ID owned by `owner`, if it exists, and 0 otherwise
    function tokenOf(address owner)
        public
        view
        virtual
        override
        returns (bytes32)
    {
        require(owner != address(0), "Invalid owner at zero address");

        return _tokens[owner];
    }

    // Returns the owner of a given token ID, reverts if the token does not exist
    function ownerOf(bytes32 tokenId)
        public
        view
        virtual
        override
        returns (address)
    {
        require(tokenId != 0, "Invalid tokenId value");

        address owner = _owners[tokenId];

        require(owner != address(0), "Invalid owner at zero address");

        return owner;
    }

    // Checks if a token ID exists
    function _exists(bytes32 tokenId) internal view virtual returns (bool) {
        return _owners[tokenId] != address(0);
    }

    // @dev Mints `tokenId` and transfers it to `to`.
    function _mint(address to, bytes32 tokenId) internal virtual {
        require(to != address(0), "Invalid owner at zero address");
        require(tokenId != 0, "Token ID cannot be zero");
        require(!_exists(tokenId), "Token already minted");
        require(tokenOf(to) == 0, "Owner already has a token");

        _tokens[to] = tokenId;
        _owners[tokenId] = to;

        emit Minted(to, tokenId, block.timestamp);
    }

    // @dev Burns `tokenId`.
    function _burn(bytes32 tokenId) internal virtual {
        address owner = Badge.ownerOf(tokenId);

        delete _tokens[owner];
        delete _owners[tokenId];

        emit Burned(owner, tokenId, block.timestamp);
    }
}

```

**Comment by cryptoregtech** [comment]
It is standard still live?

**Comment by htadashi** [comment]
With the recent NFT boom, I think this standard could make a comeback.  Besides @ra-phael use case, here is another recent example: 

https://blog.questbook.app/2021/05/29/introducing-the-decentralized-learning-economy-whitepaper-draft/#:~:text=We%20introduce%20a%20new%20class%20of%20NFT%20called%20NFT-Ls.%20These%20are%20Non%20Fungible%20Tokens%20of%20Learning.%20These%20differ%20from%20the%20standard%20NFTs%20in%20the%20sense%20that%20they%20are%20non%20tradable.%20Once%20you%20earned%20it%2C%20it%20is%20yours%20forever.

**Comment by sandys** [comment]
Goldfinch is already doing this. It has issued non-transferable NFT via off-chain verification.

https://medium.com/goldfinch-fi/introducing-unique-identity-uid-the-first-nft-for-identity-830a89207509

**Comment by vongohren** [comment]
<img width="450" alt="Screenshot 2021-11-10 at 13 40 25" src="https://user-images.githubusercontent.com/1012055/141114898-52613f15-bd48-4632-9642-f9c981b11b48.png"> This was the feedback on such an ID from the SSI community, who do identity thourougly and using blockchain in a feasable manner

**Comment by mark-smartseal** [comment]
> <img alt="Screenshot 2021-11-10 at 13 40 25" width="450" src="https://user-images.githubusercontent.com/1012055/141114898-52613f15-bd48-4632-9642-f9c981b11b48.png">
> 
> This was the feedback on such an ID as goldfinch, from the SSI community, who do identity thourougly and using blockchain in a feasable manner

Would love to learn more about their objections. Seems to me that what Goldfinch is doing really works and isn't crazy as long as the tokens are filtered correctly on the application layer and some kind of standard metadata format is used.
How does the SSI community do KYC/identity, reputation, credit, credentials, badges, etc.. for the persona associated with a wallet?

**Comment by vongohren** [comment]
I have sent you a message, but for the rest of the thread. Seek info here: https://identity.foundation/.
That will lead you to: https://www.w3.org/TR/did-core/ && https://www.w3.org/TR/vc-data-model/

VC datamodel is credentials with a signature. That again can be reputation, credit, credentials, bagdes and so on. It can also be KYC credentials. Its just a standardized credential data model that is being embraced by some of the biggest players such as Microsoft, Square and so on. It is a way to let data flow on the internet without third party intermdiarries, and the DID is what allows for cryptography to work efficiently.

**Comment by mark-smartseal** [comment]
> I have sent you a message, but for the rest of the thread. Seek info here: https://identity.foundation/. That will lead you to: https://www.w3.org/TR/did-core/ && https://www.w3.org/TR/vc-data-model/
> 
> VC datamodel is credentials with a signature. That again can be reputation, credit, credentials, bagdes and so on. It can also be KYC credentials. Its just a standardized credential data model that is being embraced by some of the biggest players such as Microsoft, Square and so on. It is a way to let data flow on the internet without third party intermdiarries, and the DID is what allows for cryptography to work efficiently.

Makes a lot of sense. A VC is a clean, gas-free way to issue a "badge". However, the revocation of that credential isn't really possible unless another separate signed message supersedes and "revokes" the previous one... which can be awkward to implement in some applications. The main benefit of issuing a VC as a token is that the VC is stored on-chain and it's easy to revoke. Revoking is important when a wallet becomes compromised. The benefit of keeping it on-chain and connected to a contract is that you don't need identity hubs. If an application is issuing positive and negative VCs against a wallet, it's better to have them connected to a single contract so they can be aggregated to form a score. If you're doing this on an identity hub then you need to trust that the hub is reporting ALL of the VCs (positive and negative) issued against that wallet in order to create an accurate aggregate score.

**Comment by vongohren** [comment]
A VC is so much more than "bagdes". You have to think broader. A VC is primarily just a datamodel that is cryptographical verifiable and semantically proovable. That is powerful.

Revocation is very much possible with simple mechanisms such as blockchain CLRs. As you can see here: https://www.w3.org/TR/vc-data-model/#example-a-holder-presenting-a-verifiable-credential-that-was-passed-to-it-by-the-subject. If you use CLR(https://en.wikipedia.org/wiki/Certificate_revocation_list) style revocation you also keep privacy intact.

You shall NEVER store credentials on the chain, that is a big privacy breach! The owner of a credential shall have this stored in an encrypted manner and present when needed.

You have to thing physical credentials to really grasp the value here. And all the same time keeping it secure and private, and dont allow for correlation by bad actors.

If you get this then you will be seeing the power of the current standards

**Comment by nikhsha** [comment]
This seems to have broader applicability for use cases such as loyalty cards. The issuer of loyalty points often doesn't want those points to be traded - they want to reward actual loyal customers by incentivising them back to their business rather than by paying them in pseudo-cash. So a balance of loyalty points that stick to a particular wallet and can only be augmented or spent with the issuer's permission seems like a pretty useful addition to the NFTs-as-loyalty-card concepts that are being kicked around.

**Comment by crjones** [comment]
Interesting post that seems relevant to this conversation:

https://vitalik.ca/general/2022/01/26/soulbound.html

**Comment by ra-phael** [comment]
From the example use cases presented in the description of this EIP, one thing becomes clear: 

**Badges can’t be bought, they must be earned.**

Badges are markers of recognition and claims about a specific identity. If someone is simply able to buy a badge, the whole reputation system falls apart. 
One shouldn’t wonder “Has this account really gained all these reputation points / karma from this DAO or they just bought them from members?”.
Additionally, if an entity wants to give special access of a service to a list of people (through accounts they control) via tokens, these should not be transferrable; otherwise a completely different set of accounts might get in.

As opposed to standards like ERC721, this standard, which makes transferring tokens from one account to another impossible, would provide a strong signal about badge ownership upon which reputation systems can be built.

One of the main criticisms against badges is the fact that making token transfers impossible is pointless since someone can just sell their account with the badges assigned to it.

- This is a valid point but maybe we could keep this standard as much simple as possible and consider this to be outside of its scope. As a result, ERC1238 would provide the guarantee that a badge can’t be transferred from an address to another, but wouldn’t claim to provide a binding between a person/entity and an address. Such continuity could be verified off-chain by services that want to leverage badges.
- At Violet we are building a toolchain to compose, express, and verify identity on-chain and this is actually a challenge that we are aware of and working on. Our first solution requires an additional layer of trust linked with the ability for one to prove they own their wallet (think 2FA using an OTP or biometrical information) to bring a higher level of assurance that the same person is still in control of an address. This is where the off-chain part comes in.
- In practice, I think that buying an externally owner account for its badges is not that attractive. In the case of a Web 2 account being sold, the first thing the buyer does is changing the email and password associated to it. However the private key associated with an EOA can’t be changed obviously. So technically the buyer doesn’t have a guarantee the seller won’t keep using that account. If the buyer could transfer the badges, they would send them to a trusted account and gave up the one bought but here they wouldn’t be able to.

It’s still work in progress but we are exploring some proposals for this EIP, starting with a naive implementation heavily inspired from ERC1155 and would love to get some feedback on it.

As mentioned in the requirements, badges can represent a unique item or a quantity, so the idea is to stay agnostic to the fungibility of badges while taking advantage of the same features brought by ERC1155, mainly managing multiple “token collections” by deploying just one contract and batch operations to mint both fungible and non-fungible tokens at once.

Here’s the interface:
[https://github.com/violetprotocol/ERC1238-token/blob/main/contracts/ERC1238/IERC1238.sol](https://github.com/violetprotocol/ERC1238-token/blob/main/contracts/ERC1238/IERC1238.sol)

Here’s an implementation:
[https://github.com/violetprotocol/ERC1238-token/blob/main/contracts/ERC1238/ERC1238.sol](https://github.com/violetprotocol/ERC1238-token/blob/main/contracts/ERC1238/ERC1238.sol)

There’s also an alternative interface and implementation using a struct to represent a `tokenCollection` which is a set of either fungible or non-fungible tokens:
[https://github.com/violetprotocol/ERC1238-token/blob/variant-with-struct/contracts/ERC1238/ERC1238.sol](https://github.com/violetprotocol/ERC1238-token/blob/variant-with-struct/contracts/ERC1238/ERC1238.sol)

The repository also contains some proposals as extensions for:

- [Staking](https://github.com/violetprotocol/ERC1238-token/pull/5)
- [“Semi-fungible”](https://github.com/violetprotocol/ERC1238-token/pull/4) tokens: Non-transferable NFTs with a shared `baseId` embedded in their `id`. For example this lets you create certificates for a course where each certificate has a unique id but one is able to lookup if an account holds a certificate for this course.

For expiration, there can be some logic in the implementer’s smart contract to burn tokens after some time has passed or burning can be externally triggered by the issuer for example.

**Comment by ccarnino** [comment]
Maybe I am naive, but in a world where we will have high TPS rollups and high data availability (cheap transactions and storage), we may want to start to store on-chain actions we do on a future web3 social app. So we can create an app with right-to-leave. 

For example a tweet or a comment on a piece of content in a web3 social app, should be soul-bound to the author. 
Why not using a Non Transferrable NFT to represent that? Let's say we have a web3 Instagram. A new photo entry should be represented as an NFT (multimedia on IPFS), while the comments and likes should be represented as Non Transferrable NFT.

This use case has different considerations and requirements compared to a badge or an earned certificate. A comment is not earned, but it's created. 

Therefore, should't the Non Transferable NFT standard be a very simple and neutral standard?

**Comment by jaygopalan** [comment]
Hey @ra-phael, thanks for working on this.

> * This is a valid point but maybe we could keep this standard as much simple as possible and consider this to be outside of its scope. As a result, ERC1238 would provide the guarantee that a badge can’t be transferred from an address to another, but wouldn’t claim to provide a binding between a person/entity and an address. Such continuity could be verified off-chain by services that want to leverage badges.

I think this is actually better. Different off-chain use cases will approach binding between the entity and the address in different ways based on tradeoffs between privacy and security (like @ccarnino said some items should be soulbound some transferrable). But all of those use cases will rely on the basic ERC1238 functionality that badges can't be transferred between addresses.

In most use cases, verifying entities in the real world probably becomes more important as the entity address accumulates more badges but err on privacy when getting started.

> * [“Semi-fungible”](https://github.com/violetprotocol/ERC1238-token/pull/4) tokens: Non-transferable NFTs with a shared `baseId` embedded in their `id`. For example this lets you create certificates for a course where each certificate has a unique id but one is able to lookup if an account holds a certificate for this course.

What do you see as different use cases for escrow vs burnable approach to staking?

**Comment by vongohren** [comment]
> This seems to have broader applicability for use cases such as loyalty cards. The issuer of loyalty points often doesn't want those points to be traded - they want to reward actual loyal customers by incentivising them back to their business rather than by paying them in pseudo-cash. So a balance of loyalty points that stick to a particular wallet and can only be augmented or spent with the issuer's permission seems like a pretty useful addition to the NFTs-as-loyalty-card concepts that are being kicked around.

How about Drivers License that is important to keep private, prescriptions from the doctor, account verification from the bank and so on.

Il stop pushing this VC mindset here because it seems that this thread wants to solve this with the possibility of on chain reputability, while I believe that too much on chain data can harm us in a bad way.

On chain is a great trust anchor, but need to architect it in a way that it does not leak data

**Comment by ccarnino** [comment]
With this standard I would focus on the core issue of "limiting minting and transferability”. Then private non transferrable NFTs, could be a future standard.

From what I read in this thread, there are different use cases that can be covered:
- non transferrable NFTs
- NFTs that can only be awarder or given
- NFTs transferrable only to other wallets I own
- NFTs transferrable if the granting authority permission
 
 We can call this new types of tokens in a lot of different ways, depending on the different combinations of needs. Let’s call them Badge Tokens for now, so they can be a flexible and composable concept.
 
 A Badge Token standard will have an interface that answers the following questions: 
 - can it only be minted by the authority (contract owner)?
  - can it be transferred to another address? 
  - has the authority’s permission to transfer to a given address?

IMO, by having an interface which replies to these questions, we can achieve a lot of different use cases, while being easily understandable.

**Comment by rosspeili** [comment]
NTTs could be also used as ban cards or restrictive flags. Eg. send an expiring NTT card to an actor with bad behavior in a virtual domain. When he tries to log in next time to the domain, eg. cryptovoxels, he is instantly transferred to the "jail" section, or just sees a black screen for the time he holds this NTT that he also can't get rid of.

Now, a little bit of imagination, and this can get us to a nightmare worst than black mirror and modern china combined, where citizen wallets or web3 IDs would get NTTs that restrict them from enter specific public spaces, museums, universities, or even travel, yet, I don't think there's a way to go around this.

Thoughts?
happy to philosophize on future use-cases on tw or dc: @rosspeili @rosspeili#0100

**Comment by Max-Levitskiy** [comment]
@rosspeili when we develop technologies, we can't create a guarantee it won't be used for some negative stuff. Kitchen knives or sweat floors in the bathroom kill a lot of people. :)
If we have some use case, which people need they will implement it. They will write a smart contract, use other blockchain or just create an off-chain solution. So, it's not a philosophical question, but a question about standardization, which is needed for better Ethereum adoption.

**Comment by nicola** [comment]
Leaving this as a note based on conversations that I had in the past few days.

Things that need to be addressed:
* Find a mechanism to get rid of NTT from a wallet (e.g. imagine someone airdropping unwanted tokens)
* Find a mechanism for tokens to be put at stake - however if lost they become burned instead of transferred.
  * I expect that this will have to do with the NTT registry implementation rather than the NTT standard
  * The use case we are targeting are putting "reputation" at risk type of dynamics.

**Comment by zhongeric** [comment]
> In practice, I think that buying an externally owner account for its badges is not that attractive. In the case of a Web 2 account being sold, the first thing the buyer does is changing the email and password associated to it. However the private key associated with an EOA can’t be changed obviously.

@ra-phael This parallel you drew to web 2 was really great, and I think we should expand on this further. I also do not think NTTs should offer any guarantee of user-address matching, but rather leave it up to services to corroborate them with their own off-chain data to verify a user. 

And on the topic of transferability (especially between wallets of the same user), looking to web2 implementations of this might be our best bet. For example, if Alice completes an IT certification course under the email **E1**, for short, a certificate is now tied to that email address in a database somewhere. Now, if Alice decides to switch to a more secure email, **E2**, she would have a few options for getting her hard earned certificate to be tied to her new account:
1. Request a new certificate from the issuer, proving she previously received it using her account under **E1**. This would lead to a _burning_ of the certificate from **E1**. The issuer can use any off-chain PII they collected to verify that Alice is the owner of both accounts (kyc, etc.) 
2. Stop using her old account under **E1** and redo the course to earn another certificate under **E2**.

Alice expects that she _will_ be able to transfer her previous certificates to her new account, but understands that she has to prove her identity beforehand to the issuer. If she doesn't want to, option 2 is less ideal but works. How can we correlate this with NTTs? 

A proposal that implements the following would be interesting to explore:
- NTTs cannot be bought.
- NTTs cannot be transferred.
- NTTs can be burned by their current holders.
- NTTs can be revoked by their issuers.

This could address the concerns about "account flipping" since the only way to reclaim an NTT would be to prove to the issuer that you previously owned it. Issuers are incentivized to prevent against sybil attacks since they would severely weaken the credibility of their issued NTTs (i.e. identity theft to reprint concert tickets). Also, users could remove unwanted NTTs from their account (malicious airdrops, attempts to give reputation / dox, etc.).

Curious to hear thoughts!

**Comment by csuwildcat** [comment]
Are folks here aware of Decentralized Identifiers and Verifiable Credentials? I ask because those are existing international standards that already provide this functionality. You can use these tools today via countless implementations and libraries available in all major languages across all platforms. The neat part is they are faster, cheaper, and more efficient for these use cases.

**Comment by vongohren** [comment]
@csuwildcat thank for stepping in and showing support for that, I have been linking to that world pretty heavily in my comments above, but its difficult to look out of the ethereum world

**Comment by wwwmaster1** [comment]
There is some great thought here, and in the DID/VC world that can be leveraged as @vongohren has been saying, but I believe there is an assumption that credentials must come from an authority - but that's not how it works IRL. Credentials come from individuals (I tell you I went to "Harvard") but the verification comes from an authority (Harvard's Records). There is obviously incentive for me to lie and there may be no cost to do so, but anyone should be able to validate it with some effort (a bit of gas), and it may not be worth ruining my reputation to do so. 

The problem I see with DID and SSI solutions these days is that there is always some third party that runs the show, or requires a special wallet, so it's not really decentralized and trustless. Why can't I create my own identity credentials (true or imaginary) and store them (or identifiers to IPFS files) in my metamask wallet, and allow them to be verified (or not) by some authority (like SSL). In fact, some of that data may even be encrypted (private medical records), and require the certification from authority in order to be decrypted.

**Comment by vongohren** [comment]
So what I read here have some troubles to me, when it comes to understanding of the SSI/DID/VC space.
There dont have to be any authority, just an issuer, can be a person. The problem you sit with is just discovery and do I trust this issuer(person), and how do one make that trust.

This is no different from a random ethereum address giving a token.

When it comes to software consuming this, you say it is a third party running the show, is not Metamask a third party running the show? They have software you install, and probably run all your transactions through Infura, which lets them decide what goes through and not? So I dont see the difference?

There is also no problem for Metamask to expand to controlling this type of software inside their wallet and send and exchange. We do see that some wallets of the DID/SSI/VC space also want to handle crypto.

Your vision is ripe to be developed, and I know there are people trying out DIDs on IPFS, encrypted storage on different mediums. So everything you talk about is possible.

There is no reason DID/SSI/VC space, should be more centralized than metamask, but Microsoft f.ex, are afraid to build it that way. But others can!

**Comment by csuwildcat** [comment]
> There is some great thought here, and in the DID/VC world that can be leveraged as @vongohren has been saying, but I believe there is an assumption that credentials must come from an authority - but that's not how it works IRL. Credentials come from individuals (I tell you I went to "Harvard") but the verification comes from an authority (Harvard's Records). There is obviously incentive for me to lie and there may be no cost to do so, but anyone should be able to validate it with some effort (a bit of gas), and it may not be worth ruining my reputation to do so.
> 
> The problem I see with DID and SSI solutions these days is that there is always some third party that runs the show, or requires a special wallet, so it's not really decentralized and trustless. Why can't I create my own identity credentials (true or imaginary) and store them (or identifiers to IPFS files) in my metamask wallet, and allow them to be verified (or not) by some authority (like SSL). In fact, some of that data may even be encrypted (private medical records), and require the certification from authority in order to be decrypted.

@wwwmaster1 what you said here belies a fundamental misunderstanding of DIDs and VCs. Verifiable Credentials can be issued by any DID (or group) to another DID, which could be a company or organization to an individual, any individual to another individual, an individual to a machine, etc. The notion that DIDs and VCs are somehow only for authorities like governments, have any binding to centralized infra, and are not decentralized is just flat out false - seriously, that's not true by any stretch of the imagination. Just because a DMV is a relatively centralized org that issues driver's licenses that people will recognize doesn't mean your DID, their DID, or Verifiable Credentials are a centralized system. You can use these decentralized components any way you want, just like some government could use a public blockchain address/contract to facilitate centralized activities that it codifies.

I would really encourage you to dig in deeper, because the risk here is that folks are going to do a ton of work recreating a slower, less efficient, more expensive version of a decentralized system that already exists, backed by international standards that are already being adopted.

**Comment by TimDaub** [comment]
Hi everyone,

I see that the discussion for non-transferrable tokens has been going on for far longer than when Vitalik wrote the Soulbound token blog post. I'd like to draw the author's and contributor's attention to another potential use case that a chapter/working group from the Radical Exchange movement is currently working on, namely: Harberger Properties or "Partial common ownership".

A few links: 

- https://partialcommonownership.com/
- https://www.radicalxchange.org/concepts/partial-common-ownership/

In one of our conversations at ETHDenver, we reached an understanding that when trying to "Harberger tax" a regular ERC721, then some of its significant functionality would have to be disabled. It is as only a specially built "common ownership" contract should ever have possession of an ERC721 token as everything else would equate to private and non-foreclosure-able property. Following this conversation, I've thought about it some more and summarized everything in a blog post that outlines the full context:

- https://timdaub.github.io/2022/02/19/non-skeuomorphic-harberger-properties-erc721-nfts/

So now here's the group's current situation. We have two main implementations going:

- https://github.com/rugpullindex/libharberger (mine that doesn't use NFTs yet)
- https://github.com/721labs/partial-common-ownership (that is a fork of @simondlr's this art work is always on sale)

In 721labs's version, you can see that the `transfer` functions are reverting to avoid private-property-type possession: https://github.com/721labs/partial-common-ownership/blob/38145d7dc4bdff772ccae016d54efe1b754b857c/contracts/token/PartialCommonOwnership.sol#L102

However, since there is arguably an interest in soulbound tokens and since Harberger Properties are somewhat "soulbound to the public" (but can change ownership rights under specific on-chain conditions), I think a way forward with this proposal may be where instead of disabling the transfer functionality, we allow implementers to configure the transfer functionality to external conditions.

So I'm proposing that we build an NFT with conditional transfers where for certain property the transfer is permanently disabled (hence the item is soulbound) and where for other types of property a transfer is only allowed when a certain condition is met.

Is this a good idea?

**Comment by zhongeric** [comment]
Why must we correlate NTTs with implementations of self-sovereign identity? It seems like we're trying to appropriate the NFT spec for a use case that it's not really meant for. The only link I see is the ability to interact with media, which could be personal information uploaded on-chain, which the consensus is strongly against. DID / VC implementations can be implemented in other fashions: [ERC725](https://github.com/ethereum/EIPs/issues/725), etc.

I no longer believe in the identity use case I previously outlined for NTTs in this [thread](https://github.com/ethereum/EIPs/issues/1238#issuecomment-1038641635), but I'm still an advocate for a simplistic standard: one that holds strict invariants preventing transferability (even in the edge case of wallets owned by the same user) and allowing both the issuing contract and owner of the NTT to arbitrarily burn them. These can exist independently of sovereign identity implementations.

**Comment by TimDaub** [comment]
I have just added my proposal for how ERC1238 could be implemented as a PR: https://github.com/ethereum/EIPs/pull/4966 Feedback is welcome.

**Comment by RetroWorld415** [comment]
How can I use the soulbound token or the concept of Soulbound? I want to creatively use it for rewarding people for contributing to the project. Where can I get a test code or code for it? I would like to experiment with it.

**Comment by MicahZoltu** [comment]
There is some discussion about this over at Ethereum Magicians: https://ethereum-magicians.org/t/eip-4973-account-bound-tokens

There is no standard yet though.

**Comment by AdedamolaXL** [comment]
Can NTT be applied to encrypting data/media such that only those in possession of the token can view it ?

**Comment by davidfurlong** [comment]
> Can NTT be applied to encrypting data/media such that only those in possession of the token can view it ?

if you encrypted the data individually with each public key of a wallet issued an NTT, and then maybe stored the link/encrypted data in the NTT metadata it could work.

**Comment by ra-phael** [comment]
@AdedamolaXL Yes, this is possible using [Lit Protocol](https://litprotocol.com/) for example.

**Comment by kamilk91** [comment]
Guys, can you explain me one thing?

So, lets make a story.:

I have an account on forum "Ethereum is cool forum"
Im earning experience points and it for example gives me some roles, or idk, colors of nick

Forum's contract needs to send me just 1x NTT and its updating it with experience?
Or im receiving nev ntt when im gaining 1exp point
or maybe there is some standalone mechanizm of experience, that gives me NTT depends on exp points level?

**Comment by MicahZoltu** [comment]
Closing this issue for housekeeping purposes. People are welcome to continue discussing in this thread, but for additional visibility an EIP should be created or the conversation should be migrated to https://ethereum-magicians.org/

**Comment by maprob** [comment]
cant you just add a require statement in the transfer function

**Comment by LucazFFz** [comment]
If the standard requires implementers to allow holders to burn their tokens, why is a burn function not included in the specification interface?

**Comment by Pandapip1** [comment]
This is not actually a standard. If @nicola or would like to actually make this a standard, I would be willing to help. However, there are already many existing EIPs that do this:

https://eips.ethereum.org/EIPS/eip-5192
https://eips.ethereum.org/EIPS/eip-5484
https://eips.ethereum.org/EIPS/eip-5114
https://eips.ethereum.org/EIPS/eip-5516
https://eips.ethereum.org/EIPS/eip-5633
https://eips.ethereum.org/EIPS/eip-5727
https://eips.ethereum.org/EIPS/eip-6454
